modules/io/common/src/main/java/org/locationtech/jts/io/twkb/TWKBHeader.java

```diff
@@ -0,0 +1,277 @@
+        POINT(1, GeometryFactory::createPoint), //
+        LINESTRING(2, GeometryFactory::createLineString), //
+        POLYGON(3, GeometryFactory::createPolygon), //
+        MULTIPOINT(4, GeometryFactory::createMultiPoint), //
+        MULTILINESTRING(5, GeometryFactory::createMultiLineString), //
+        MULTIPOLYGON(6, GeometryFactory::createMultiPolygon), //
+        GEOMETRYCOLLECTION(7, GeometryFactory::createGeometryCollection);
+
+        private final int value;
+
+        private final Function<GeometryFactory, Geometry> emptyBuilder;
+
+        GeometryType(int value, Function<GeometryFactory, Geometry> emptyBuilder) {
+            this.value = value;
+            this.emptyBuilder = emptyBuilder;
+        }
+
+        public int getValue() {
+            return value;
+        }
+
+        // held as a class variable cause calling values() on a tight loop has a non-depreciable
+        // performance impact
+        private static final GeometryType[] VALUES = GeometryType.values();
+
+        public static GeometryType valueOf(int value) {
+            return VALUES[value - 1];
+        }
+
+        public static GeometryType valueOf(Class<? extends Geometry> gclass) {
+            Objects.requireNonNull(gclass);
+            if (Point.class.isAssignableFrom(gclass))
+                return POINT;
+            if (LineString.class.isAssignableFrom(gclass))
+                return LINESTRING;
+            if (Polygon.class.isAssignableFrom(gclass))
+                return POLYGON;
+            if (MultiPoint.class.isAssignableFrom(gclass))
+                return MULTIPOINT;
+            if (MultiLineString.class.isAssignableFrom(gclass))
+                return MULTILINESTRING;
+            if (MultiPolygon.class.isAssignableFrom(gclass))
+                return MULTIPOLYGON;
+            if (GeometryCollection.class.isAssignableFrom(gclass))
+                return GEOMETRYCOLLECTION;
+
+            throw new IllegalArgumentException("Unrecognized geometry tpye: " + gclass);
+        }
+
+        public Geometry createEmpty(GeometryFactory factory) {
+            return this.emptyBuilder.apply(factory);
+        }
+    }
+
+    private GeometryType geometryType;
+
+    private int xyPrecision = 0;
+
+    private boolean hasBBOX = false;
+
+    private boolean hasSize = false;
+
+    private boolean hasIdList = false;
+
+    public boolean hasExtendedPrecision() {
+        return hasZ() || hasM();
+    }
+
+    private boolean isEmpty = false;
+
+    private boolean hasZ = false;
+
+    private boolean hasM = false;
+
+    private int zPrecision = 0;
+
+    private int mPrecision = 0;
+
+    /**
+     * Size of encoded geometry body, iif size_flag == 1, defaults to {@code -1} if {@link #hasSize}
+     * ({@code == false}
+     * <p>
+     * {@code geometry_body_size := uint32 # size in bytes of <geometry_body>}
+     */
+    private int geometryBodySize;
+
+    /////////////////////// custom optimizations ///////////////////////
```

Comment:
How are these optimizations?

Reply:
TLTR: this comment is obsolete, and I've now removed it.

@groldan branch included some optimizations that could be enabled, or disabled in order to be compliant with PostGIS version from 2019.
Since PostGIS has included some of the optimizations (https://trac.osgeo.org/postgis/ticket/4461), thus I've changed JTS encoder implementation to be aligned with PostGIS one, and removed the enabling/disabling switch from encoder API.

Reply:
Interesting... @dr-jts do you have any input on that optimization in PostGIS?

While it makes sense to skip over repeated points, I wonder if there are topological considerations.  Those considerations could be either direction!

I suppose a good question would be: "Is the code both compliant with the standard and useful with PostGIS?" 

If JTS can be made to work with both, I'm super happy!:)